### ISSUE TITLE
Idea/Discussion: Porting partition module to QML

### ISSUE THREAD

**Issue by AIIX** [calamares/calamares#1954]
#### **Is your feature request related to a problem? Please describe.**

Hi, I have been working on customising Calamares for the Plasma Bigscreen project which requires using QML modules over Qt Widget modules to enable navigation via Arrow Keys and USB connected remotes, Currently the only module that seems to be lacking a QML equivalent seems to be the main Partition Module. 

#### **Describe the solution you'd like**

I have been working on porting the Partition Module to QML, progressing slowly as the module is quite complex and has nested functionality embedded into the gui widgets logic. 

My current approach to porting the module to QML has been to test and study how the current widget UI works and how it connects to its widgets backend, then additionally testing how much of the widget based functionality and data models can be accessed in QML, if if not creating proxy functionality that can supply the required data to QML objects.

#### How the partitionq module is designed:

- Follows a sub page design where the main partitionq qml provides a loader object to load the subpages of the partition module depending on the current process flow.
- The functionality on the cpp backend is divided as per the subpages, so a sub page called ChoosePage.qml will have a cpp backend class called ChoicePageQml that will handle all the functionality within the context of that page. Follows the similar approach to the widget module except it should house all the functions required by the page in the class itself.

#### Where the above port is at currently:

- Initializing the Partition Core Module
- Displaying the Initial install choose page with a working device selector combo box
- Displaying the install choices supported by the device
- Displaying swap choices supported when selecting the "Erase" install option
- Encryption selection and setting the encryption passphrase

#### To Do:

- Everything else, this is not a functional port yet

#### Some road blocks I have come across till now and possible solutions I see to them:

- Tree View models: The PartitionModel currently provides a tree view interface for Qt Widgets, QML Quick 2 does not have any native support for displaying Tree Views as such, Possible Solution: Add a proxy model to partitionq to flatten the Tree structure to be able to use it in QML listview or repeaters

- Nested functionality of install options: Options such as Install Alongside and Replace partition depend on accessing PartitionSplitterWidget, Possible Solution: Copy the minimal needed functionality from the widget UI to perform the same steps in the partitionq module, this might also be applicable to more parts of the module because of the nested functionality but the idea is probably going to be the same is try copying the minimal amount of functionality needed for it work with the QML port

**Comment by demmm** [comment]
Thanks for posting this.  I have been working off & on for about a year to port the last module used in KaOS not ready for QML, which is ofc the partition module (good to hear all other modules you need are ported).
My setup so far is also to use Loader to start the different options, but absolute nothing working yet, just a general GUI idea.
I will try your work, and see what fits.

**Comment by kkofler** [comment]
> for the Plasma Bigscreen project which requires using QML modules over Qt Widget modules to enable navigation via Arrow Keys and USB connected remotes

Is not that the real issue you want to solve?

I believe we could have a lot more convergence-friendly applications if it did not require rewriting all applications in QML. IMHO, QtWidgets needs to be made to handle remote and touch input properly rather than shelved in favor of QML. (Touch input by itself actually works, but there is a lack in QtWidgets of touch-friendly widgets like the ones Kirigami provides for QML. Remote control support seems actually more broken, judging from your description.)

**Comment by AIIX** [comment]
> > for the Plasma Bigscreen project which requires using QML modules over Qt Widget modules to enable navigation via Arrow Keys and USB connected remotes
> 
> Is not that the real issue you want to solve?
> 

I don't see fixing Qt Widgets key navigation as an easy solution not for me at least, instead of just porting that one module to QML when Calamares already supports loading QML modules with the exception of partition module.

But I don't disagree that it would have just been less messy and much easier if Qt Widgets did support Key Navigation in a way that is expected to work with TV and USB remote input which QML does allow to do quite easily, but also fixing Qt Widgets is a task way beyond my current capabilities as I am not well versed with the Qt Widgets, so if Qt or someone in the ecosystem would fix it, I would be happy to use Qt Widgets, if not then there isn't much of a option but to try and get this one module ported and working with QML (which I am more versed with) if I want to be able to use Calamares in Plasma Bigscreen project.

**Comment by demmm** [comment]
@AIIX One connection is now made thanks to your fork (Calamares builds & runs with partitionq), the deviceModel correctly sets the 2 available drives (text from the model not displayed yet.
Just an image of what I am working on:
![Screenshot_20220519_222925](https://user-images.githubusercontent.com/5360485/169399738-713198ce-661f-4554-bf30-f73b12f9b998.png)

**Comment by AIIX** [comment]
> @AIIX One connection is now made thanks to your fork (Calamares builds & runs with partitionq), the deviceModel correctly sets the 2 available drives (text from the model not displayed yet. Just an image of what I am working on: 

The UI looks really good and seems much cleaner then the UI I have currently working in my branch, is the code for the UI hosted somewhere ? I think it would be great to use this UI since its looks already more polished and complete.

**Comment by demmm** [comment]
No, nothing uploaded yet, since the C++ code used is yours.  I can add it to the KaOS fork, but since the module builds, runs & doesn't crash Calamares (plus it doesn't effect the existing partition module at all), it might be best to get the partitionq module up in Calamares here.
Question for @adriaandegroot really, work in forks, or move it here?

For anyone following, a few more shots of what partitonq might look like.  As @AIIX already mentioned, the TableView for the disk layout won't be easy to port to QML, so I just created a ColumnLayout with hardcoded coded colors & numbers to get the UI (visible once a partition option is selected).  UI uses elements from Mauikit as is now. Second shots shows the hover effect for the tooltips.

![Screenshot_20220520_114044](https://user-images.githubusercontent.com/5360485/169502962-4245faaa-6fcb-4e2c-9f2a-e33e360eeca5.png)
![Screenshot_20220520_114124](https://user-images.githubusercontent.com/5360485/169502970-10ef1ed9-1c15-4172-9aea-abe651380d4a.png)

**Comment by AIIX** [comment]
> No, nothing uploaded yet, since the C++ code used is yours. I can add it to the KaOS fork, but since the module builds, runs & doesn't crash Calamares (plus it doesn't effect the existing partition module at all), it might be best to get the partitionq module up in Calamares here. Question for @adriaandegroot really, work in forks, or move it here?

+1 it would be nice if we can get a branch here in Calamares itself, even if its not ready to be in master so everyone intested can work on a common branch (and then merge it to master when its more ready) than keeping it in the separate forks

**Comment by adriaandegroot** [comment]
@demmm @AIIX I've invited AiiX to R/W status to the repo, so then you can work together in a straightorward way in a feature branch. Demmm has full developer status so can then merge things as needed.

**Comment by demmm** [comment]
@AIIX partitionq branch is created:
https://github.com/calamares/calamares/tree/partitionq
Best that you get your work in first.

**Comment by AIIX** [comment]
@demmm I have created a PR for it on https://github.com/calamares/calamares/pull/1960

**Comment by demmm** [comment]
@AIIX I have not opened a PR for the QML files I have, since they would overwrite your work.  I did try to work from yours first, but partitionq will not load with those here, crashes right away. My files are just a GUI concept, with virtually no connections yet (only connected is showing the devices), so probably not a good idea to have those in the partitonq branch.
